sqlglot/dataframe/sql/dataframe.py

```diff
@@ -0,0 +1,741 @@
+        if with_expression:
+            existing_ctes = with_expression.args["expressions"]
+            existsing_cte_names = [x.alias_or_name for x in existing_ctes]
+            for cte in ctes:
+                if cte.alias_or_name not in existsing_cte_names:
+                    existing_ctes.append(cte)
+        else:
+            existing_ctes = ctes
+        expression.set("with", exp.With(expressions=existing_ctes))
+        return expression
+
+    @classmethod
+    def _get_outer_select_columns(cls, item: t.Union[exp.Expression, DataFrame]) -> t.List[Column]:
+        expression = item.expression if isinstance(item, DataFrame) else item
+        return [Column(x) for x in dict.fromkeys(expression.find(exp.Select).args.get("expressions", []))]
+
+    @classmethod
+    def _create_hash_from_expression(cls, expression: exp.Select):
+        value = expression.sql(dialect="spark").encode("utf-8")
+        return f"t{zlib.crc32(value)}"[:6]
+
+    def _get_select_expressions(self) -> t.List[t.Tuple[t.Union[exp.Cache, OutputExpressionContainer], exp.Select]]:
+        select_expressions = []
+        main_select_ctes = []
+        for cte in self.expression.ctes:
+            cache_storage_level = cte.args.get("cache_storage_level")
+            if cache_storage_level:
+                select_expression = cte.this.copy()
+                select_expression.set("with", exp.With(expressions=copy(main_select_ctes)))
+                select_expression.set("cte_alias_name", cte.alias_or_name)
+                select_expression.set("cache_storage_level", cache_storage_level)
+                select_expressions.append((exp.Cache, select_expression))
+            else:
+                main_select_ctes.append(cte)
+        main_select = self.expression.copy()
+        if main_select_ctes:
+            main_select.set("with", exp.With(expressions=main_select_ctes))
+        select_expressions.append((type(self.output_expression_container), main_select))
+        return select_expressions
+
+    def sql(self, dialect="spark", optimize=True, **kwargs) -> t.List[str]:
+        df = self._resolve_pending_hints()
+        select_expressions = df._get_select_expressions()
+        output_expressions = []
+        replacement_mapping = {}
+        for expression_type, select_expression in select_expressions:
+            select_expression = select_expression.transform(replace_id_value, replacement_mapping)
+            if optimize:
+                select_expression = optimize_func(select_expression)
+            select_expression = df._replace_cte_names_with_hashes(select_expression)
+
+            # temporary hack to remove extra CTE that was replaced with table until removing unused CTEs is added
+            # to optimizer
+            filtered_ctes = [
+                cte
+                for cte in select_expression.ctes
+                if cte.alias_or_name not in {x.alias_or_name for x in replacement_mapping.values()}
```

Comment:
replaceme_mapping double for loop?

Reply:
Yeah this is part of the hacky section that will be replaced once we improve the optimizer to remove unused CTEs.